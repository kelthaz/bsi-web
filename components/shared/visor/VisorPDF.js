import { useEffect, useRef, useState } from 'react';
import { Document, Page, pdfjs } from 'react-pdf';
import SimuladorRepositorio from '../../../services/simulador/simulador.repositorio';
import styles from './visor-pdf.module.scss';

const VisorPDF = ({ data, extension = 'pdf', container }) => {
  pdfjs.GlobalWorkerOptions.workerSrc = `//cdnjs.cloudflare.com/ajax/libs/pdf.js/${pdfjs.version}/pdf.worker.min.js`;

  const pageRef = useRef();
  const [numPages, setNumPages] = useState(null);
  const [pageNumber, setPageNumber] = useState(1);
  const [scale, setScale] = useState(1);
  const [degree, setDegree] = useState(0);
  const [expand, setExpand] = useState(false);

  const [url, setUrl] = useState('');

  useEffect(async () => {
    const tablaPdf = await SimuladorRepositorio.postSimuladorTablaPdf({
      monto: 6200000,
      periodicidad: 2,
      plazo: 18,
    }).then((resp) => resp);
    // console.log(tablaPdf.data);
    const urlBlob = window.URL.createObjectURL(new Blob([tablaPdf.data], { extension }));
    setUrl(urlBlob);
  }, []);

  const onDocumentLoadSuccess = ({ numPages: nPages }) => setNumPages(nPages);

  const onPageLoadSuccess = (page) => {
    pageRef.current = page;
    console.log(pageRef.current);
  };

  const changePage = (offset) => setPageNumber((prevPageNumber) => prevPageNumber + offset);
  const changeDegree = (degreesAdd) =>
    setDegree((prevDegreesAdd) =>
      degreesAdd + prevDegreesAdd >= 360 ? degreesAdd + prevDegreesAdd - 360 : degreesAdd + prevDegreesAdd
    );

  const previousPage = () => changePage(-1);

  const nextPage = () => changePage(1);

  const turnRight = () => changeDegree(90);

  const turnLeft = () => changeDegree(270);

  const onZoomIn = () => setScale(scale * 1.1);
  const onZoomOut = () => setScale(scale * 0.9);

  const calculateScale = () => {
    const { clientWidth, clientHeight } = container.current;
    const { originalWidth, originalHeight } = pageRef.current;

    const rotated = degree % 180 !== 0;
    const pageWidth = rotated ? originalHeight : originalWidth;
    const pageHeight = rotated ? originalWidth : originalHeight;
    const scaleX = clientWidth / pageWidth;
    const scaleY = clientHeight / pageHeight;

    setExpand(!expand);
    setScale(expand ? scale * scaleX : scale * Math.min(scaleX, scaleY));
  };

  console.log(degree % 180 !== 0 ? 'rotate' : 'no rotate');
  console.log(pageRef.current);

  return (
    <div className={styles['contenedor-pdf']}>
      <Document file="/Hunk-160R-FI-Hero-MotoCorp.pdf" onLoadSuccess={onDocumentLoadSuccess}>
        <Page pageNumber={pageNumber} scale={scale} rotate={degree} onLoadSuccess={onPageLoadSuccess} ref={pageRef} />
      </Document>

      <div>
        <div className="pagec">
          Page {pageNumber || (numPages ? 1 : '--')} of {numPages || '--'}
        </div>
        <div className="buttonc">
          <button type="button" disabled={pageNumber <= 1} onClick={previousPage} className="Pre">
            Previous
          </button>
          <button type="button" disabled={pageNumber >= numPages} onClick={nextPage}>
            Next
          </button>
          <button type="button" disabled={pageNumber >= numPages} onClick={onZoomIn}>
            +
          </button>
          <button type="button" disabled={pageNumber >= numPages} onClick={onZoomOut}>
            -
          </button>
          <button type="button" disabled={pageNumber >= numPages} onClick={calculateScale}>
            expand
          </button>
          <button type="button" disabled={pageNumber >= numPages} onClick={turnLeft}>
            izquierda
          </button>
          <button type="button" disabled={pageNumber >= numPages} onClick={turnRight}>
            derecha
          </button>
        </div>
      </div>
    </div>
  );
};

export default VisorPDF;
