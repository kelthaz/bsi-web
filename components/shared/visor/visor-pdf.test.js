import '@testing-library/jest-dom';
import { mount, shallow } from 'enzyme';
import React, { useEffect } from 'react';
import VisorPDF from './VisorPDF';

jest.mock('react-pdf', () => ({
  pdfjs: { GlobalWorkerOptions: { workerSrc: 'abc' } },
  Document: () => <div />,
  Outline: null,
  Page: () => <div />,
}));

describe('Pruebas en el componente VisorPDF', () => {
  const container = {
    current: {
      clientWidth: 450,
    },
  };
  const pageDimentions = { originalWidth: 350, originalHeight: 300 };
  const numPages = { numPages: 4 };
  test('Debe de mostrarse correctamente el componente', async () => {
    // arrange
    const wrapper = shallow(<VisorPDF container={container} />);
    // act
    const { onLoadSuccess } = wrapper.find('Document').props();
    onLoadSuccess(numPages);
    // assert
    expect(wrapper).toMatchSnapshot();
  });

  test('Debe de avanzar a la siguiente pagina', () => {
    // arrange
    const wrapper = shallow(<VisorPDF container={container} />);
    const buttonNextjs = wrapper.find('button').at(2);
    // act
    buttonNextjs.simulate('click');
    const pageNumber = wrapper.find('span').text().split('/')[0].trim();
    // assert
    expect(pageNumber).toBe('2');
  });

  test('Debe de avanzar 2 paginas y luego regresar 1 pagina', () => {
    // arrange
    const wrapper = shallow(<VisorPDF container={container} />);
    const buttonNextjs = wrapper.find('button').at(2);
    const previusNextjs = wrapper.find('button').at(1);
    // act
    buttonNextjs.simulate('click');
    buttonNextjs.simulate('click');
    previusNextjs.simulate('click');

    const pageNumber = wrapper.find('span').text().split('/')[0].trim();
    // assert
    expect(pageNumber).toBe('2');
  });

  test('Debe de rotar 2 veces el pdf hacia la izquierda', () => {
    // arrange
    const wrapper = shallow(<VisorPDF container={container} />);
    const turnLeft = wrapper.find('button').at(3);
    // act
    turnLeft.simulate('click');
    turnLeft.simulate('click');
    const { rotate } = wrapper.find('Page').props();
    // assert
    expect(rotate).toBe(180);
  });

  test('Debe de rotar 2 veces el pdf hacia la derecha', () => {
    // arrange
    const wrapper = shallow(<VisorPDF container={container} />);
    const turnRight = wrapper.find('button').at(4);
    // act
    turnRight.simulate('click');
    turnRight.simulate('click');
    const { rotate } = wrapper.find('Page').props();
    // assert
    expect(rotate).toBe(180);
  });

  test('Debe de acercar el pdf', () => {
    // arrange
    const wrapper = shallow(<VisorPDF container={container} />);
    const { onLoadSuccess } = wrapper.find('Page').props();
    onLoadSuccess(pageDimentions);
    const onZoomIn = wrapper.find('button').at(5);
    // act
    onZoomIn.simulate('click');
    const { scale } = wrapper.find('Page').props();
    // assert
    expect(scale).toBe(1 * 1.1);
  });

  test('Debe de de alejar el pdf', () => {
    // arrange
    const wrapper = shallow(<VisorPDF container={container} />);
    const { onLoadSuccess } = wrapper.find('Page').props();
    onLoadSuccess(pageDimentions);
    const onZoomOut = wrapper.find('button').at(6);
    // act
    onZoomOut.simulate('click');
    const { scale } = wrapper.find('Page').props();
    // assert
    expect(scale).toBe(1 * 0.9);
  });
  test('Debe de de expandir el pdf', () => {
    // arrange
    const wrapper = shallow(<VisorPDF container={container} />);
    const { onLoadSuccess } = wrapper.find('Page').props();
    onLoadSuccess(pageDimentions);
    const turnLeft = wrapper.find('button').at(3);
    turnLeft.simulate('click');
    // act
    const expand = wrapper.find('button').at(7);
    expand.simulate('click');
    const { scale } = wrapper.find('Page').props();
    // assert
    expect(scale).toBe(1 * (container.current.clientWidth / pageDimentions.originalHeight));
  });
  test('Debe restaurar el pdf', () => {
    // arrange
    const wrapper = shallow(<VisorPDF container={container} />);
    const { onLoadSuccess } = wrapper.find('Page').props();
    onLoadSuccess(pageDimentions);
    // act
    wrapper.find('button').at(7).simulate('click');
    wrapper.find('button').at(7).simulate('click');
    const { scale } = wrapper.find('Page').props();
    // assert
    expect(scale).toBe(1);
  });
});
