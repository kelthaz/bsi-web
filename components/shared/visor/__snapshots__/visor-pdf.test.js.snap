// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente VisorPDF Debe de mostrarse correctamente el componente 1`] = `
<Fragment>
  <div
    className="container-visor-pdf "
  >
    <Document
      file="/Hunk-160R-FI-Hero-MotoCorp.pdf"
      onLoadSuccess={[Function]}
    >
      <Page
        onLoadSuccess={[Function]}
        pageNumber={1}
        rotate={0}
        scale={1}
      />
    </Document>
  </div>
  <div
    className="container-controls-visor-pdf"
  >
    <div
      className="buttons-controls-visor-pdf"
    >
      <button
        type="button"
      >
        <SvgDownload />
      </button>
      <button
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        <SvgArrow />
      </button>
      <button
        className="turn-arrow"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <SvgArrow />
      </button>
      <button
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <SvgTurn />
      </button>
      <span>
        1
         / 
        4
      </span>
      <button
        className="turn-right"
        onClick={[Function]}
        type="button"
      >
        <SvgTurn />
      </button>
      <button
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <SvgPlus />
      </button>
      <button
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <SvgMinus />
      </button>
      <button
        className=""
        onClick={[Function]}
        type="button"
      >
        <SvgExpand />
      </button>
    </div>
  </div>
</Fragment>
`;
